src/wallet/test/wallet_tests.cpp

```diff
@@ -334,6 +334,82 @@ BOOST_AUTO_TEST_CASE(LoadReceiveRequests)
     BOOST_CHECK_EQUAL(values[1], "val_rr1");
 }
 
+// Test some watch-only wallet methods by the procedure of loading (LoadWatchOnly),
+// checking (HaveWatchOnly), getting (GetWatchPubKey) and removing (RemoveWatchOnly) a
+// given PubKey, resp. its corresponding P2PK Script. Results of the the impact on
+// the address -> PubKey map is dependent on whether the PubKey is cryptographically
+// correct or not
+static void TestWatchOnlyPubKey(CWallet& wallet, const CPubKey& add_pubkey)
+{
+    CScript p2pk = GetScriptForRawPubKey(add_pubkey);
+    CKeyID add_address = add_pubkey.GetID();
+    CPubKey found_pubkey;
+
+    // all Scripts (i.e. also all PubKeys) are add to the general watch-only set
+    BOOST_CHECK(!wallet.HaveWatchOnly(p2pk));
+    wallet.LoadWatchOnly(p2pk);
+    BOOST_CHECK(wallet.HaveWatchOnly(p2pk));
+
+    // only cryptographically valid PubKeys shall be add to the watch-only address -> PubKey map
+    bool isPubKeyFullyValid = add_pubkey.IsFullyValid();
+    if (isPubKeyFullyValid) {
+        BOOST_CHECK(wallet.GetWatchPubKey(add_address, found_pubkey));
+        BOOST_CHECK(found_pubkey == add_pubkey);
+    } else {
+        BOOST_CHECK(!wallet.GetWatchPubKey(add_address, found_pubkey));
+    }
+
+    wallet.RemoveWatchOnly(p2pk);
+    BOOST_CHECK(!wallet.HaveWatchOnly(p2pk));
+
+    if (isPubKeyFullyValid)
+        BOOST_CHECK(!wallet.GetWatchPubKey(add_address, found_pubkey));
+}
+
+// Cryptographically invalidate a PubKey whilst keeping length and first byte
+static void PollutePubKey(CPubKey& pubkey)
+{
+    do {
+        std::vector<unsigned char> pubkey_raw(pubkey.begin(), pubkey.end());
+        std::random_shuffle(pubkey_raw.begin()+1, pubkey_raw.end());
```

Comment:
Would it just be simpler for this test to zero out the last 32/64 bytes? Is having the shuffled bytes testing anything important
?

Reply:
Good point -- no, the shuffling doesn't test anything import, for some reason I didn't want to set the PubKey coordinate(s) to a fixed value off the curve. Zeroing out is simpler and saves the loop.